test/decode-next-page-query-test.js

```diff
@@ -454,4 +454,82 @@ suite('decode-next-page-query-test', () => {
       }
     `));
   });
+
+  test('it retains variable values on multiple next-page-queries', () => {
+    const variables = [variable('first', 'Int', 1)];
+    const cursor = 'product-cursor';
+    const fragmentFixture = {
+      data: {
+        shop: {
+          products: {
+            pageInfo: {
+              hasNextPage: true,
+              hasPreviousPage: false
+            },
+            edges: [{
+              cursor,
+              node: {
+                id: productId
+              }
+            }]
+          }
+        }
+      }
+    };
+    const document = new Document(typeBundle);
+
+    document.addQuery(variables, (root) => {
+      root.add('shop', (shop) => {
+        shop.addConnection('products', {args: {first: variables[0]}}, (products) => {
+          products.add('id');
+        });
+      });
+    });
+
+    const decodedPageOne = decode(document.operations[0], fragmentFixture.data);
+
+    const [pageTwoQuery] = decodedPageOne.shop.products[0].nextPageQueryAndPath();
+
+    assert.deepEqual(tokens(pageTwoQuery.toString()), tokens(`
+      query ($first: Int = 1) {
+        shop {
+          products (first: $first, after: "${cursor}") {
```

Comment:
so before this would have been `first: 1` and then the query would be invalid because the `$first` variable was unused?

Reply:
Initial query would have been `products(first: 1)`, then we hoist to a var and go `product(first: $first, after: "cursor")`, and then on the next iteration we do the same thing and set the value of the new var to the value in the arg (which is now a var) and it fails at `query($first: Int = $first)`.

It gets an extra iteration of success when we trim the the tree and query off the nearest node, so other integration tests in js buy and manual testing didn't catch this.

Reply:
makes sense 👍

Reply:
The hard part about syntax being flexible is that you have to account for syntax being flexible. blargh.